bitcoin/src/psbt/mod.rs

```diff
@@ -72,87 +290,457 @@ impl Psbt {
+            (Some(_), Some(locktime)) => Ok(locktime),
+            _ => {
+                if let Some(locktime) = inner.fallback_locktime {
+                    return Ok(locktime);
+                }
+                Ok(absolute::LockTime::ZERO)
+            }
+        }
+    }
+
+    /// Generates the unique transaction ID for this [`Psbt`].
+    /// See https://bips.xyz/370#unique-identification.
+    ///
+    /// ## Errors
+    ///
+    /// - [`Error::RequiredLocktimeNotPresent`] - If this is a PSBTv2 and
+    /// the existing inputs don't have a compatible locktime.
+    fn get_unique_id(&self) -> Result<Txid, Error> {
+        let mut unsigned_tx = match self.inner.version {
+            Version::PsbtV0 => self.inner.unsigned_tx.as_ref().unwrap().clone(),
+            _ => self.construct_unsigned_tx()?,
         };
-        psbt.unsigned_tx_checks()?;
-        Ok(psbt)
+
+        // BIP 370: Since PSBT_IN_SEQUENCE can be changed by Updaters and Combiners,
+        //          the sequence number in this unsigned transaction must be set to 0
+        //          (not final, nor the sequence in PSBT_IN_SEQUENCE).
+        for txin in unsigned_tx.input.iter_mut() {
+            txin.sequence = Sequence::ZERO;
+        }
+
+        Ok(unsigned_tx.txid())
+    }
+
+    /// Constructs a new unsigned transaction from this Psbt.
+    /// Should be used for Non-PsbtV0s only. Don't use this function
+    /// to compare the PSBTs when PSBTv2s are involved, instead use
+    /// the `Psbt::get_unique_id()` function.
+    fn construct_unsigned_tx(&self) -> Result<Transaction, Error> {
+        if self.inner.version == Version::PsbtV0 {
+            return Ok(self.inner.unsigned_tx.as_ref().unwrap().clone());
+        }
+        let mut tx = Transaction {
+            version: self.inner.tx_version.unwrap(),
+            lock_time: self.compute_locktime()?,
+            input: vec![],
+            output: vec![],
+        };
+
+        for psbtin in self.inner.inputs.iter() {
+            tx.input.push(TxIn {
+                previous_output: OutPoint {
+                    txid: psbtin.previous_tx_id.unwrap(),
+                    vout: psbtin.output_index.unwrap(),
+                },
+                script_sig: ScriptBuf::new(),
+                sequence: psbtin.sequence.unwrap_or_default(),
+                witness: Witness::default(),
+            });
+        }
+
+        for psbtout in self.inner.outputs.iter() {
+            tx.output.push(TxOut {
+                value: psbtout.amount.unwrap(),
+                script_pubkey: psbtout.script.as_ref().unwrap().clone(),
+            });
+        }
+
+        Ok(tx)
     }
 
-    /// Extracts the `Transaction` from a PSBT by filling in the available signature information.
-    pub fn extract_tx(self) -> Transaction {
-        let mut tx: Transaction = self.unsigned_tx;
+    /// Converts this Psbt into a PsbtV0
+    pub fn get_v0(self) -> Result<Self, Error> {
```

Comment:
These `get_vN` functions returning `Self` are smelly to me. It's almost like `Psbt` is a generic on the Version without making that distinction explicit using the type system.

It might make more sense to have  `Psbt` be `PsbtV2` as inner since it's strictly more capable and can be serialized to v0 at any time.

The `PSBT` type would have ergonomic and validating interfaces like you've described here, while `PsbtV0` and `PsbtV2` types similar to current `PsbtInner` could allow manual field manipulation. There are occasions in the `payjoin` library downstream where protocol dictates [PSBTv0 manual field manipulation](https://github.com/payjoin/rust-payjoin/blob/d0cde6176b8ce09c18ea5f9336f687cd253becd4/payjoin/src/receive/mod.rs#L773-L809), so exposing fields seems to me a requirement for current feature parity of an upgrade.

Reply:
Thanks for the detailed information, will dig into it more.